gr-analog/lib/agc3_cc_impl.cc

```diff
@@ -56,81 +56,137 @@ agc3_cc_impl::agc3_cc_impl(float attack_rate,
 agc3_cc_impl::~agc3_cc_impl() {}
 
 
+float agc3_cc_impl::attack_rate() const
+{
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    return d_attack;
+}
+float agc3_cc_impl::decay_rate() const
+{
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    return d_decay;
+}
+float agc3_cc_impl::reference() const
+{
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    return d_reference;
+}
+float agc3_cc_impl::gain() const
+{
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    return d_gain;
+}
+float agc3_cc_impl::max_gain() const
+{
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    return d_max_gain;
+}
+
+void agc3_cc_impl::set_attack_rate(float rate)
+{
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    d_attack = rate;
+}
+void agc3_cc_impl::set_decay_rate(float rate)
+{
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    d_decay = rate;
+}
+void agc3_cc_impl::set_reference(float reference)
+{
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    d_reference = reference;
+}
+void agc3_cc_impl::set_gain(float gain)
+{
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    d_gain = gain;
+    d_gain_change = true;
+}
+void agc3_cc_impl::set_max_gain(float max_gain)
+{
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    d_max_gain = max_gain;
+}
+
 int agc3_cc_impl::work(int noutput_items,
                        gr_vector_const_void_star& input_items,
                        gr_vector_void_star& output_items)
 {
-    const gr_complex* in = (const gr_complex*)input_items[0];
-    gr_complex* out = (gr_complex*)output_items[0];
-
-#ifdef __GNUC__
-    // Compute a linear average on reset (no expected)
-    if (__builtin_expect(d_reset, false)) {
-        float mags[noutput_items] __attribute__((aligned(16)));
-        volk_32fc_magnitude_32f(mags, &in[0], noutput_items);
-#else
-    // Compute a linear average on reset (no expected)
-    if (!d_reset) {
-        _declspec(align(16)) std::vector<float> mags(noutput_items);
-        volk_32fc_magnitude_32f(&mags[0], &in[0], noutput_items);
-#endif
-        float mag(0.0);
-        for (int i = 0; i < noutput_items; i++) {
-            mag += mags[i];
+    auto in = reinterpret_cast<const gr_complex*>(input_items[0]);
+    auto out = reinterpret_cast<gr_complex*>(output_items[0]);
+    float attack;
+    float decay;
+    float reference;
+    float gain;
+    float max_gain;
+    unsigned int iir_update_decim;
+    // make copies of the runtime-adjustable floats
+    // we do this so we can be finely granular when locking
+    {
+        gr::thread::scoped_lock guard(d_setter_mutex);
+        attack = d_attack;
+        decay = d_decay;
+        reference = d_reference;
+        gain = d_gain;
+        max_gain = d_max_gain;
+        iir_update_decim = d_iir_update_decim;
+        d_gain_change = false; // if gain setting happened, we've just used the new gain.
+    }
```

Comment:
Is there really much to be gained by reducing the scope of locking? The requisite `d_gain_change` makes the code more complex.

Reply:
probably not, you're right. let me uncruftify my code.